### PR TITLE
[1205] FhirEngine Sync Download API Re-design

### LINKED ISSUE #1205: FhirEngine Sync Download API Re-design
**Is your feature request related to a problem? Please describe.**
Modify the Android FhirEngine Download APIs to give developers more control over which[ FHIR resources](https://www.hl7.org/fhir/resourcelist.html) to sync, and how to store these resources on the phone

**Describe the solution you'd like**
An implementation that gives developers control on exactly which FHIR resources can be synced to the phone. Developers can write the app to fetch resources via search parameters, or IDs. Developers can constrain the resources downloaded to a subset of FHIR resources. The user-supplied functions give app developers power to tailor their code to specify which resources to fetch.

**Describe alternatives you've considered**
Current implementation 

**Additional context**
See: https://docs.google.com/document/d/1UZ7Ndb8JRmB2UdXpfIkPT-VAra6X2aLSzdGI_a8KJkg

**Would you like to work on the issue?**
Yes

### PR DESCRIPTION
**Pull request by omarismail94**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1205 

**Description**
FhirEngine Sync Download API Redesign

See for details: https://docs.google.com/document/d/1UZ7Ndb8JRmB2UdXpfIkPT-VAra6X2aLSzdGI_a8KJkg/edit#

TESTED=all unit tests passed. Added one more unit test as well

tl;dr: give more functionality to developers to control what resources they can download.

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?
No, open to suggestions though!

**Type**
Choose one: (Bug fix | **Feature** | Documentation | Testing | Code health | Builds | Releases | Other)

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

File: engine/src/main/java/com/google/android/fhir/sync/FhirSyncWorker.kt [hunk 1]
```diff
@@ -35,22 +35,34 @@ import kotlinx.coroutines.cancel
...
 /** A WorkManager Worker that handles periodic sync. */
 abstract class FhirSyncWorker(appContext: Context, workerParams: WorkerParameters) :
   CoroutineWorker(appContext, workerParams) {
   abstract fun getFhirEngine(): FhirEngine
   abstract fun getDataSource(): DataSource
-  abstract fun getSyncData(): ResourceSyncParams
+  abstract fun getInitialUrl(): String
+  abstract fun getCreateDownloadUrl(): (String, String?) -> String
+  abstract fun getExtractResourcesFromResponse(): (Resource) -> Collection<Resource>
+  abstract fun getExtractNextUrlsFromResource(): (Resource) -> Collection<String>
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/sync/FhirSyncWorker.kt:49]: Should we encapsulate this in something like a `SyncDownloadContext`? Engine can provide a default implementation for it based on what we have implemented in `FhirPeriodicSyncWorker` and Developer  applications  may use it or provide their custom implementation based on the use case?

**Comment by Tarun-Bhardwaj** [comment]
@omarismail94 , did you get a chance to work on the review comments? This issue is lined up for FHIR Engine Beta release next week.
CC: @jingtang10

File: engine/src/main/java/com/google/android/fhir/sync/FhirSyncWorker.kt (hunk 1 above)
Comment (omarismail94) [reply engine/src/main/java/com/google/android/fhir/sync/FhirSyncWorker.kt]: Great idea! I created an interface named `SyncDownloadExtractor` that users can implement. Provided a sample implementation in the `demo` code, This interface is need to instantiate a `FhirSyncWorker`

**Comment by omarismail94** [comment]
@aditya-07 lmk if there are other comments that need addressing. I created a new interface named `SyncDownloadExtractor` that users can implement. An implementation of this interface is needed to instantiate a `FhirSyncWorker`. 

Note that I refactored the functions to take in parameters now as opposed to creating lambdas to make things cleaner


> @omarismail94 , did you get a chance to work on the review comments? This issue is lined up for FHIR Engine Beta release next week. CC: @jingtang10

@Tarun-Bhardwaj  yes :)

File: demo/src/main/java/com/google/android/fhir/demo/data/SyncDownloadExtractorImpl.kt [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    if (lastUpdate != null && downloadUrl.contains("\$everything")) {
+      downloadUrl = "$downloadUrl?_since=$lastUpdate"
+    }
+
+    if (lastUpdate != null && !downloadUrl.contains("\$everything")) {
+      downloadUrl = "$downloadUrl&_lastUpdated=gt$lastUpdate"
+    }
+
+    if (downloadUrl.contains("&page_token")) {
+      downloadUrl = preProcessUrl
```
Comment (jingtang10) [inline demo/src/main/java/com/google/android/fhir/demo/data/SyncDownloadExtractorImpl.kt:52]: some documentation / comments would be great for this

File: demo/src/main/java/com/google/android/fhir/demo/data/SyncDownloadExtractorImpl.kt [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      }
+    }
+
+    if (resourceResponse is Bundle) {
+      val nextUrl =
+        resourceResponse.link.firstOrNull { component -> component.relation == "next" }?.url
+      if (nextUrl != null) {
+        queueWork.add(nextUrl)
+      }
+    }
```
Comment (jingtang10) [inline demo/src/main/java/com/google/android/fhir/demo/data/SyncDownloadExtractorImpl.kt:83]: add a brief comment to explain the 2 cases

File: engine/src/test/java/com/google/android/fhir/sync/SyncJobTest.kt [hunk 4]
```diff
@@ -146,7 +148,7 @@ class SyncJobTest {
     val flow = MutableSharedFlow<State>()
     val job = launch { flow.collect { res.add(it) } }
 
-    syncJob.run(fhirEngine, dataSource, resourceSyncParam, flow)
+    syncJob.run(fhirEngine, dataSource, TestingUtils.TestSyncDownloadExtractorImpl, flow)
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/sync/SyncJobTest.kt:151]: the `resourceSyncParam` in this class can be deleted?

**Review by jingtang10 — CHANGES_REQUESTED** [review]
are there no tests to collect the flow of the resources? do you think we should add a test class for fhirsynchronizer?

File: engine/src/main/java/com/google/android/fhir/sync/SyncDownloadExtractor.kt [hunk 5]
```diff
@@ -16,17 +16,12 @@
...
 
-import android.content.Context
-import androidx.work.WorkerParameters
-import com.google.android.fhir.resource.TestingUtils
-import org.hl7.fhir.r4.model.ResourceType
+import org.hl7.fhir.r4.model.Resource
 
-class TestFailingSyncWorker(appContext: Context, workerParams: WorkerParameters) :
-  FhirSyncWorker(appContext, workerParams) {
+interface SyncDownloadExtractor {
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/sync/SyncDownloadExtractor.kt:21]: just brainstorming, this is optional :)
```suggestion
interface DownloadProcessor {
```


```suggestion
interface DownloadHandler {
```

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: engine/src/main/java/com/google/android/fhir/sync/FhirSynchronizer.kt [hunk 6]
```diff
@@ -102,14 +103,35 @@ internal class FhirSynchronizer(
...
-        downloadResourceType(it.key, it.value)
-      } catch (exception: Exception) {
-        exceptions.add(ResourceSyncException(it.key, exception))
+    fhirEngine.syncDownload {
+      flow {
+        val listOfUrls = mutableListOf<String>()
+        listOfUrls.add(syncDownloadExtractor.getInitialUrl())
+        while (listOfUrls.isNotEmpty()) {
+          try {
+            val preprocessedUrl = listOfUrls.removeAt(0)
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/FhirSynchronizer.kt:116]: Better make `listOfUrls` a `Queue<String>` and use appropriate apis.

[the remark this answers is not in the document]
Comment (omarismail94) [reply engine/src/main/java/com/google/android/fhir/sync/FhirSynchronizer.kt]: Changed to a queue. Good idea!

File: engine/src/main/java/com/google/android/fhir/sync/SyncDownloadExtractor.kt (hunk 5 above)
Comment (omarismail94) [reply engine/src/main/java/com/google/android/fhir/sync/SyncDownloadExtractor.kt]: I saw the `FhirSynchronizer` contains a reference to `Uploader`, so I changed the name to `Downloader`. Lmk if that works!

**Comment by omarismail94** [comment]
> are there no tests to collect the flow of the resources? do you think we should add a test class for fhirsynchronizer?

 @jingtang10  Added a test to count the number of times resources get loaded that should be equal to the length of the queue

Also addressed all comments you made. Lmk if there's anything else you want!

File: engine/src/main/java/com/google/android/fhir/sync/SyncDownloadExtractor.kt (hunk 5 above)
Comment (jingtang10) [reply engine/src/main/java/com/google/android/fhir/sync/SyncDownloadExtractor.kt]: i think the uploader does something different, it actually does the uploading!

i think here maybe download processor is better. @aditya-07 even suggested DownloadPrePostProcessor, or DownloadManager?

**Review by jingtang10 — APPROVED** [review]
thanks @omarismail94!

File: engine/src/main/java/com/google/android/fhir/sync/SyncDownloadExtractor.kt (hunk 5 above)
Comment (omarismail94) [reply engine/src/main/java/com/google/android/fhir/sync/SyncDownloadExtractor.kt]: Renamed to `DownloadManager`!
